Fancy preorder reasoning

ohad opened the pull request:
Change the poor-man's preorder reasoning in `contrib` to use slightly faster data structures, following [Agda](https://github.com/agda/agda-stdlib/issues/622).

Since only the `TailRec` ursive version of `Data.List` seems to use these proofs, hopefully it'll be ok to change this now.

gallais commented on libs/contrib/Data/List/TailRec.idr:
```diff
@@ -19,9 +19,9 @@ module Data.List.TailRec
+import Data.List 
+import Data.Vect 
+import Data.Nat  
 
```
Trailing whitespace

gallais commented on libs/contrib/Data/List/TailRec.idr:
```diff
@@ -33,36 +33,25 @@ length xs = length_aux xs Z
+  ~~ Data.List.length xs + 0             ...( sym $ plusZeroRightNeutral $ Data.List.length xs ) 
+  ~~ Data.List.TailRec.length_aux xs 0   ...( lemma 0 xs )                                       
+  ~~ Data.List.TailRec.length xs  ...( Refl )                                             
   where
```
Should we have a notation for qed? (Also: alignment)

ohad commented:
@gallais : both syntactic comments are because I reverted `import Data.List as DL`, so once we implemented `import .. as`, these will go away (just kidding, will push a fix).

Re. `QED`: it's not needed, we have a `DEQ` in the beginning of the proof.
